pkg/member/member_control.go

```diff
@@ -136,6 +144,34 @@ func (m *memberControl) AddMemberAsLearner(ctx context.Context) error {
 	return nil
 }
 
+// AddLearnerWithRetry add a new member as a learner with exponential backoff.
+func (m *memberControl) AddLearnerWithRetry(ctx context.Context, retrySteps int, dataDir string) error {
+	backoff := miscellaneous.CreateBackoff(RetryPeriod, retrySteps)
+
+	if err := retry.OnError(backoff, func(err error) bool {
+		return err != nil
+	}, func() error {
+		// Additional safety check before adding a learner
+		if _, err := os.Stat(dataDir); err == nil {
+			if err := os.RemoveAll(filepath.Join(dataDir)); err != nil {
+				return fmt.Errorf("failed to remove directory %s with err: %v", dataDir, err)
+			}
+		} else if !os.IsNotExist(err) {
+			return err
```

Comment:
If the data directory does not exist then an error is returned. retry.OnError will retry 6 times and every time this error will be returned. I guess you want an early exit if the data directory is not present. In that case please create an error and then filter that error in `retriable` function argument to `retry.OnError`

Reply:
> I guess you want an early exit if the data directory is not present

I don't get it, what makes you think so ?
```go
 func() error {
		if (some condition) {
                         // do something
			return err --> here it will retry
		} 
		return nil ---> here it will exit
	}
```
If err is other than `data dir not exist`, then we will retry 
else 
better to try to add a learner